### ISSUE TITLE
User permissions: grant/revoke access on databases

### ISSUE THREAD

**Issue by niolap** [siodb/siodb#37]
I think there is not innovation to make here. Utlimatly, we would have the following components:

- Role: Administrator or User
  - Administrator: can do everything and read/write everything by default.
  - User: can only log in by default. Can't see the system table 'by design'.
- Group: just group, to group users
- User
- Permission: The permission that a group or a user can have. E.g.
  - read_only: permission to read only a database
  - read_write: permission to read_write a database

What we have currently:

- Role: We just have the concept of super User. Super User is the administrator role.
- User

What we need in this issue:

- Role: Be sure that non Administrator user cannot read system tables
  - Administrator: can do everything and read/write everything by default.
  - User: can only log in by default. Can't see the system table 'by design'. Thus unless provided the correct permission, a user can do nothing by default excepted creating a session.
- Permission: The permission that a user can have. 
  - read_only: permission to read only a database
  - read_write: permission to read_write a database

SQL to grant permissions to users:

```sql
alter user <user> grant read_only on database <database>;
alter user <user> grant read_write on database <database>;
alter user <user> grant read_only on database all;
alter user <user> grant read_write on database all;
```

```sql
alter user <user> revoke read_only from database <database>;
alter user <user> revoke read_write from database <database>;
alter user <user> revoke read_only from database all;
alter user <user> revoke read_write from database all;
```

We also need a table to list user permission. I think there is this table `SYS_USER_PERMISSIONS` that already has this purpose.
We don't need the concept of `GRANT_OPTIONS` here.
Maybe a kind of memory table `sys_permissions` to list available permission would be nice with comment would be nice (nice to have).

For security reasons, if a user tries to access an object for which he has no access. The message should be "object does not exist" instead of "no permission on this object".

**Comment by niolap** [comment]
@cxxman posted above the design suggestion. Ready for feedback.

**Comment by cxxman** [comment]
So what about groups? I suggest to not implement them.

**Comment by cxxman** [comment]
I think this model can be simplified a bit. `read_write`, `read_only` and `admin` can be permissions that can be granted to user. `admin` can read and update data, and also can do DDL, `read_only` can only read data, `read_write` can both read and write data but no DDL. In addition to that, user can be granted permission to create and drop databases - these two should be granted separately and `admin` should not include this.

**Comment by niolap** [comment]
@cxxman I like this simplification. But, from my point of view, there is an instance administrators and database users. There is no administrator per database. So I would sum up everything like this:

- `read_write`: can (Query + DML + DDL) only in a database (`on database <database>`) or all databases (`on database all`)
  - The goal is to have a quick permission that allow user to to work completely within his database
- `read_only`: can Query only in a database (`on database <database>`) or all databases (`on database all`)
- `admin`: can do everything in all databases (`on database all`). `root` is `admin` be design (not revokable).
- `create_database`: can create database and has automatically `read_write` permission on its created database.
  - By default, a created user can only create a session. He must be granted the permission to use a database or to create one explicitly by a user having an `admin` permission.

In the future, we may add other permissions like `ddl`, `dml` and so on and maybe classical `create table`, `select table`, `update table` etc. But no such need now.

Don't you think?

**Comment by cxxman** [comment]
I think we need to have more fine-grained permissions like `select table` underneath, otherwise it will be significantly harder to add them later, but do not expose them now into SQL. SQL will have those higher level things like `read_only` or `read_write`, which will map into an appropriate set of those fine-grained permissions.

**Comment by niolap** [comment]
> So what about groups? I suggest to not implement them.

Yes. Not needed now.

**Comment by niolap** [comment]
> I think we need to have more fine-grained permissions like `select table` underneath, otherwise it will be significantly harder to add them later, but do not expose them now into SQL. SQL will have those higher-level things like `read_only` or `read_write`, which will map into an appropriate set of those fine-grained permissions.

100% Ok.

**Comment by niolap** [comment]
@cxxman one point come up to my mind and I just would like to be sure, when we grant/revoke permission to/from a database object, this must include all existing and **future** objects. Meaning that if I create a table `T` in database `D` after having granted permission `READ_ONLY` for user `U`, then user `U` will be able to select `T` without having to execute a second grant command. Correct?

**Comment by cxxman** [comment]
@niolap For this we need to implement special form of the GRANT statement, like this:
```sql
-- for the specified database
GRANT permissions ON database_name.* TO user_name

-- for a currently selected database
GRANT permissions ON * TO user_name
```
Notice that star instead of table name.

**Comment by niolap** [comment]
> @niolap For this we need to implement special form of the GRANT statement, like this:
> 
> ```sql
> -- for the specified database
> GRANT permissions ON database_name.* TO user_name
> 
> -- for a currently selected database
> GRANT permissions ON * TO user_name
> ```
> 
> Notice that star instead of the table name.

@cxxman Yes. I think this is the best way to implement this. Then, it must be included in this issue because we'll need it significantly for our project.

**Comment by cxxman** [comment]
@niolap Don't we also need some SQL command which will show current permissions granted to user?
